File: ML_week_03/mot/src/mobile/android/app/src/main/java/com/example/mot/TrackerEngine.kt
package com.example.mot

import android.content.Context
import android.graphics.RectF
import androidx.camera.core.ImageProxy
import org.tensorflow.lite.Interpreter
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel

/**
 * TrackerEngine – TFLite inference wrapper for Android.
 *
 * Loads `siamese_tracker.tflite` from the assets folder and exposes
 * [initializeTracking] + [track] for per-frame Siamese tracking.
 */
class TrackerEngine(context: Context) {

    companion object {
        private const val MODEL_FILE = "siamese_tracker.tflite"
        private const val TEMPLATE_SIZE = 127
        private const val SEARCH_SIZE = 255
    }

    private var interpreter: Interpreter? = null
    private var currentBBox = RectF()

    // Pre-allocated input buffers
    private val templateBuffer: ByteBuffer =
        ByteBuffer.allocateDirect(4 * TEMPLATE_SIZE * TEMPLATE_SIZE * 3).apply {
            order(ByteOrder.nativeOrder())
        }
    private val searchBuffer: ByteBuffer =
        ByteBuffer.allocateDirect(4 * SEARCH_SIZE * SEARCH_SIZE * 3).apply {
            order(ByteOrder.nativeOrder())
        }

    init {
        val model = loadModelFile(context)
        val options = Interpreter.Options().apply {
            setNumThreads(4)
            // Uncomment for GPU: addDelegate(GpuDelegate())
        }
        interpreter = Interpreter(model, options)
    }

    // ── Public API ───────────────────────────────────────────────

    /** Initialise tracking with first-frame bbox (image coordinates). */
    fun initializeTracking(frame: ImageProxy, bbox: RectF) {
        currentBBox = RectF(bbox)
        // 1. Crop template patch from frame centred on bbox
        // 2. Resize to TEMPLATE_SIZE × TEMPLATE_SIZE
        // 3. Write normalised RGB floats into templateBuffer
    }

    /** Track target in current frame. Returns updated bbox. */
    fun track(frame: ImageProxy): RectF {
        // 1. Crop search region from frame around currentBBox
        // 2. Resize to SEARCH_SIZE × SEARCH_SIZE
        // 3. Write into searchBuffer
        // 4. Run interpreter
        // 5. Read response map, find peak, update currentBBox
        val inputs = arrayOf(templateBuffer, searchBuffer)
        // Output shape depends on model — adjust accordingly
        val responseMap = Array(1) { Array(17) { FloatArray(17) } }
        // interpreter?.runForMultipleInputsOutputs(inputs, mapOf(0 to responseMap))
        // ... find peak, convert to image coords, update currentBBox
        return currentBBox
    }

    fun close() {
        interpreter?.close()
        interpreter = null
    }

    // ── Helpers ──────────────────────────────────────────────────

    private fun loadModelFile(context: Context): MappedByteBuffer {
        val fd = context.assets.openFd(MODEL_FILE)
        val inputStream = FileInputStream(fd.fileDescriptor)
        val channel = inputStream.channel
        return channel.map(FileChannel.MapMode.READ_ONLY, fd.startOffset, fd.declaredLength)
    }
}

File: ML_week_03/mot/src/mobile/ios/Sources/TrackerEngine.swift
// TrackerEngine.swift – TFLite inference wrapper for iOS.
//
// Loads siamese_tracker.tflite from the app bundle and provides
// initializeTracking / track methods for per-frame object tracking.

import CoreGraphics
import CoreVideo
import Foundation

// Requires TensorFlowLiteSwift pod / SPM package
// import TensorFlowLite

final class TrackerEngine {

    // MARK: - Properties

    private let templateSize = 127
    private let searchSize = 255

    // TFLite interpreter (uncomment when TensorFlowLite is linked)
    // private var interpreter: Interpreter?

    /// Stored template embedding / patch (float buffer)
    private var templateBuffer: [Float] = []
    /// Last known bbox (x, y, w, h) in image coordinates
    private var currentBBox: CGRect = .zero

    // MARK: - Initialiser

    init() {
        guard let modelPath = Bundle.main.path(forResource: "siamese_tracker", ofType: "tflite") else {
            print("[TrackerEngine] ERROR: siamese_tracker.tflite not found in bundle.")
            return
        }
        print("[TrackerEngine] Model path: \(modelPath)")
        // interpreter = try? Interpreter(modelPath: modelPath)
        // try? interpreter?.allocateTensors()
    }

    // MARK: - Public API

    /// Initialise tracking with the first frame and user-selected bbox.
    func initializeTracking(frame: CVPixelBuffer, bbox: CGRect) {
        currentBBox = bbox
        // 1. Crop template patch from frame centred on bbox.
        // 2. Resize to (templateSize × templateSize).
        // 3. Normalise to [0,1] float32.
        // 4. Store in templateBuffer.
        print("[TrackerEngine] Initialised with bbox=\(bbox)")
    }

    /// Track the target in the given frame.
    /// - Returns: Updated bounding box in image coordinates.
    func track(frame: CVPixelBuffer) -> CGRect {
        // 1. Crop search patch centred on currentBBox (with context).
        // 2. Resize to (searchSize × searchSize).
        // 3. Normalise to [0,1] float32.
        // 4. Set template tensor + search tensor on interpreter.
        // 5. Invoke interpreter.
        // 6. Read response map output.
        // 7. Apply cosine window + find peak.
        // 8. Convert peak to new image-space bbox.
        // 9. Update currentBBox.
        return currentBBox
    }

    // MARK: - Helpers

    /// Convert CVPixelBuffer region to a normalised Float array.
    private func cropAndNormalise(
        _ buffer: CVPixelBuffer,
        roi: CGRect,
        targetSize: Int
    ) -> [Float] {
        // Lock base address, extract BGRA, convert to RGB float [0,1],
        // resize to targetSize × targetSize.
        // Return flat float array of length targetSize*targetSize*3.
        return Array(repeating: 0.0, count: targetSize * targetSize * 3)
    }
}
